### PR TITLE
Fixed AssetBrowser handling multiple files with same name.

### LINKED ISSUE #14158: Asset Browser displaying wrong files if multiple files exist with the same name.
Test case: sidec01.mix and sidec02.mix -> radar.shp.
If multiple container / packages contain a file with the same filename, the asset browser shows always the same / last one, regardless of which container / package you choose in the above select-box.
This only affects the asset browser. Sequences referencing these sprites / frames in-game can address both files and display them individualy.

### PR DESCRIPTION
**Pull request by IceReaper**
This fixes #14158.

### CODE REVIEW

**Comment by reaperrr** [comment]
This has admittedly been sitting unreviewed for far too long, but we most likely won't get this merged before the playtest, so adding to Next+1 milestone.

**Review by Mailaender — DISMISSED** [review]
Works as promised.

**Comment by reaperrr** [comment]
Might be Windows-specific, but I'm getting

> Error reading 50cal.shp:
 Illegales Zeichen im Pfad.
   bei System.IO.Path.CheckInvalidPathChars(String path, Boolean checkAdditional)
   bei System.IO.Path.Combine(String path1, String path2)
   bei OpenRA.FileSystem.Folder.Contains(String filename) in d:\dev\OpenRA\OpenRA.Game\FileSystem\Folder.cs:Zeile 50.
   bei OpenRA.FileSystem.FileSystem.<>c__DisplayClass13.<TryOpen>b__12(IReadOnlyPackage x) in d:\dev\OpenRA\OpenRA.Game\FileSystem\FileSystem.cs:Zeile 259.
   bei System.Linq.Enumerable.LastOrDefault[TSource](IEnumerable`1 source, Func`2 predicate)
   bei OpenRA.FileSystem.FileSystem.TryOpen(String filename, Stream& s) in d:\dev\OpenRA\OpenRA.Game\FileSystem\FileSystem.cs:Zeile 259.
   bei OpenRA.FileSystem.FileSystem.Open(String filename) in d:\dev\OpenRA\OpenRA.Game\FileSystem\FileSystem.cs:Zeile 212.
   bei OpenRA.Graphics.SpriteLoader.GetFrames(IReadOnlyFileSystem fileSystem, String filename, ISpriteLoader[] loaders) in d:\dev\OpenRA\OpenRA.Game\Graphics\SpriteLoader.cs:Zeile 111.
   bei OpenRA.Graphics.SpriteLoader.GetSprites(IReadOnlyFileSystem fileSystem, String filename, ISpriteLoader[] loaders, SheetBuilder sheetBuilder) in d:\dev\OpenRA\OpenRA.Game\Graphics\SpriteLoader.cs:Zeile 106.
   bei OpenRA.Graphics.SpriteCache.LoadSprite(String filename, List`1 cache) in d:\dev\OpenRA\OpenRA.Game\Graphics\SpriteLoader.cs:Zeile 61.
   bei OpenRA.Graphics.SpriteCache.get_Item(String filename) in d:\dev\OpenRA\OpenRA.Game\Graphics\SpriteLoader.cs:Zeile 73.
   bei OpenRA.Mods.Common.Widgets.Logic.AssetBrowserLogic.LoadAsset(IReadOnlyPackage package, String filename) in d:\dev\OpenRA\OpenRA.Mods.Common\Widgets\Logic\AssetBrowserLogic.cs:Zeile 343.
Error reading 120mm.shp:
 Illegales Zeichen im Pfad.
   bei System.IO.Path.CheckInvalidPathChars(String path, Boolean checkAdditional)
   bei System.IO.Path.Combine(String path1, String path2)
   bei OpenRA.FileSystem.Folder.Contains(String filename) in d:\dev\OpenRA\OpenRA.Game\FileSystem\Folder.cs:Zeile 50.
   bei OpenRA.FileSystem.FileSystem.<>c__DisplayClass13.<TryOpen>b__12(IReadOnlyPackage x) in d:\dev\OpenRA\OpenRA.Game\FileSystem\FileSystem.cs:Zeile 259.
   bei System.Linq.Enumerable.LastOrDefault[TSource](IEnumerable`1 source, Func`2 predicate)
   bei OpenRA.FileSystem.FileSystem.TryOpen(String filename, Stream& s) in d:\dev\OpenRA\OpenRA.Game\FileSystem\FileSystem.cs:Zeile 259.
   bei OpenRA.FileSystem.FileSystem.Open(String filename) in d:\dev\OpenRA\OpenRA.Game\FileSystem\FileSystem.cs:Zeile 212.
   bei OpenRA.Graphics.SpriteLoader.GetFrames(IReadOnlyFileSystem fileSystem, String filename, ISpriteLoader[] loaders) in d:\dev\OpenRA\OpenRA.Game\Graphics\SpriteLoader.cs:Zeile 111.
   bei OpenRA.Graphics.SpriteLoader.GetSprites(IReadOnlyFileSystem fileSystem, String filename, ISpriteLoader[] loaders, SheetBuilder sheetBuilder) in d:\dev\OpenRA\OpenRA.Game\Graphics\SpriteLoader.cs:Zeile 106.
   bei OpenRA.Graphics.SpriteCache.LoadSprite(String filename, List`1 cache) in d:\dev\OpenRA\OpenRA.Game\Graphics\SpriteLoader.cs:Zeile 61.
   bei OpenRA.Graphics.SpriteCache.get_Item(String filename) in d:\dev\OpenRA\OpenRA.Game\Graphics\SpriteLoader.cs:Zeile 73.
   bei OpenRA.Mods.Common.Widgets.Logic.AssetBrowserLogic.LoadAsset(IReadOnlyPackage package, String filename) in d:\dev\OpenRA\OpenRA.Mods.Common\Widgets\Logic\AssetBrowserLogic.cs:Zeile 343.
(...)

File: OpenRA.Mods.Common/Widgets/Logic/AssetBrowserLogic.cs [hunk 1]
```diff
@@ -312,27 +314,33 @@ bool LoadAsset(string filename)
...
 
 			try
 			{
+				currentPackage = package;
+				currentFilename = filename;
+				var prefix = "";
+				var fs = modData.DefaultFileSystem as OpenRA.FileSystem.FileSystem;
+
+				if (fs != null)
+					prefix = fs.GetPrefix(package) + "|";
```
Comment (RoosterDragon) [inline OpenRA.Mods.Common/Widgets/Logic/AssetBrowserLogic.cs]: This change fixes it for me.

```c#
if (fs != null)
{
	prefix = fs.GetPrefix(package);
	if (prefix != null)
		prefix += "|";
}
```

File: OpenRA.Mods.Common/Widgets/Logic/AssetBrowserLogic.cs [hunk 2]
```diff
@@ -368,17 +376,38 @@ bool ShowSourceDropdown(DropDownButtonWidget dropdown)
...
 
-			var files = assetSource != null ? assetSource.Contents : modData.ModFiles.MountedPackages.SelectMany(f => f.Contents).Distinct();
-			foreach (var file in files.OrderBy(s => s))
+			var files = new SortedList<string, List<IReadOnlyPackage>>();
+
+			if (assetSource != null)
+				foreach (var content in assetSource.Contents)
+					files.Add(content, new List<IReadOnlyPackage> {
+						assetSource
+					});
```
Comment (RoosterDragon) [inline OpenRA.Mods.Common/Widgets/Logic/AssetBrowserLogic.cs]: I think sticking the `files.Add` is fine.

File: OpenRA.Mods.Common/Widgets/Logic/AssetBrowserLogic.cs [hunk 3]
```diff
@@ -368,17 +376,38 @@ bool ShowSourceDropdown(DropDownButtonWidget dropdown)
...
+				foreach (var mountedPackage in modData.ModFiles.MountedPackages)
 				{
-					AddAsset(assetList, file, template);
-					availableShps.Add(file);
+					foreach (var content in mountedPackage.Contents)
+					{
+						if (!files.ContainsKey(content))
+							files.Add(content, new List<IReadOnlyPackage> {
+								mountedPackage
+							});
```
Comment (RoosterDragon) [inline OpenRA.Mods.Common/Widgets/Logic/AssetBrowserLogic.cs]: I think sticking the `files.Add` on one line is fine.

**Comment by RoosterDragon** [comment]
With the above fix, works for me. Style comments are optional.

**Comment by pchote** [comment]
Does someone want to take this over?

**Review by reaperrr — DISMISSED** [review]

**Comment by reaperrr** [comment]
Updated with @RoosterDragon's fix, works fine as far as I can tell.

**Review by reaperrr — DISMISSED** [review]

**Review by reaperrr — APPROVED** [review]

**Comment by reaperrr** [comment]
https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/3073b38d31008f4fce2d35edca24b558fe0d7b9f...b862e076857831e5b0d2c0b98d77beda6f1b3fd1

**Pull request merged**